source/simulator/melt.cc

```diff
@@ -1036,30 +1036,40 @@ namespace aspect
             Assert(melt_outputs != NULL, ExcMessage("Need MeltOutputs from the material model for computing the melt variables."));
             const FEValuesExtractors::Vector fluid_velocity_extractor = this->introspection().variable("fluid velocity").extractor_vector();
 
+            std::vector<Tensor<1,dim> > phi_u_f(dofs_per_cell);
+
             for (unsigned int q=0; q<n_q_points; ++q)
-              for (unsigned int i=0; i<dofs_per_cell; ++i)
-                {
-                  for (unsigned int j=0; j<dofs_per_cell; ++j)
-                    cell_matrix(i,j) += fe_values[fluid_velocity_extractor].value(j,q) *
-                                        fe_values[fluid_velocity_extractor].value(i,q) *
-                                        fe_values.JxW(q);
+              {
+                for (unsigned int i=0; i<dofs_per_cell; ++i)
+                  {
+                    phi_u_f[i] = fe_values[fluid_velocity_extractor].value(i,q);
```

Comment:
if you wanted to optimize this further, then you'd initialize the vector `phi_u_f` above by zeros and added an `if (i is a u_f DoF) to this loop. In fact, a similar statement could be made in the i-j double loop below.